knowledge-base/grid-custom-filter-menu.md

```diff
@@ -0,0 +1,201 @@
+description: Learn how to refresh filter menu context after programmatic changes with a custom component
+type: how-to
+page_title: How to Refresh Filter Menu Context After Programmatic Changes
+slug: grid-kb-custom-filter-menu
+tags: grid, blazor, custom, filter, filtermenu
+res_type: kb
+ticketid: 1669381
+---
+
+## Environment
+<table>
+    <tbody>
+	    <tr>
+	    	<td>Product</td>
+	    	<td>Grid for Blazor</td>
+	    </tr>
+    </tbody>
+</table>
+
+## Description
+
+When using the TelerikGrid in Blazor applications, it's common to implement custom `FilterMenu` for enhanced filtering capabilities. However, integrating an [AutoComplete]({%slug autocomplete-overview%}) component with `StringOperators` within a [`FilterMenuTemplate`]({%slug grid-templates-filter%}#filter-menu-template) for dynamic filtering based on user input does not work as expected. The AutoComplete component does not consider the selection from the dropdown list of `StringOperators` within the `FilterMenuTemplate`. 
+
+This knowledge base article also answers the following questions:
+- How to integrate AutoComplete with `StringOperators` in a TelerikGrid `FilterMenuTemplate`?
+- How to refresh a custom `FilterMenuTemplate` in TelerikGrid?
+- How to refresh a `FilterMenuTemplate` context after programmatic changes?
+
+## Solution
+
+To achieve the desired behavior, encapsulate the content of the `FilterMenuTemplate` in a custom component and refresh this component upon selection change in the dropdown list. Below is an example demonstrating this approach:
+
+<div class="skip-repl"></div>
+````Home.razor
+@using Telerik.DataSource
+<TelerikGrid Data="@Countries"
+             FilterMode="GridFilterMode.FilterMenu"
+             PageSize="25">
+    <GridColumns>
+        <GridColumn Field="@(nameof(Country.CountryName))">
+            <FilterMenuTemplate>
+                <CustomFilterMenu @bind-FilterOperator="@FilterOperator"
+                                  @bind-SelectedCountry="@SelectedCountry"
+                                  Countries="@Countries"
+                                  FilterOperators="@FilterOperators" />
+            </FilterMenuTemplate>
+            <FilterMenuButtonsTemplate Context="filterContext">
+                <TelerikButton OnClick="@(async _ => await FilterAsync(filterContext))" Class="k-button-solid-primary">Filter</TelerikButton>
+                <TelerikButton OnClick="@(async _ => await ClearFilterAsync(filterContext))">Clear</TelerikButton>
+            </FilterMenuButtonsTemplate>
+        </GridColumn>
+    </GridColumns>
+</TelerikGrid>
+
+@code {
+    private IEnumerable<StringFilterOperator> FilterOperators { get; set; } = Enum.GetValues(typeof(StringFilterOperator)).Cast<StringFilterOperator>().ToList();
+    private StringFilterOperator FilterOperator { get; set; } = StringFilterOperator.StartsWith;
+
+    private string SelectedCountry { get; set; } = null!;
```

Comment:
`null!` is cheating in this case, because the value is not populated all the time.

```suggestion
    private string SelectedCountry { get; set; } = string.Empty;
```